src/com/cdd/bao/template/SchemaTree.java

```diff
@@ -138,7 +149,72 @@ public String toString()
 		}
 		return ancestors.toArray(new String[ancestors.size()]);
 	}
+
+	// nodes is a list of (parentURI, Node) pairs, where each node is a candidate
+	// for this SchemaTree, and should preset label, description, and uri fields
+	// return true if SchemaTree was changed, false otherwise
+	public boolean addNodes(List<Pair<String, Node>> nodes)
+	{
+		List<Pair<String, Node>> candidates = new ArrayList<>(nodes); 
+		while (!candidates.isEmpty())
+		{
+			int cSize = candidates.size();
+			for (Iterator<Pair<String, Node>> it = candidates.iterator(); it.hasNext();)
+			{
+				Pair<String, Node> pair = it.next();
+				Node parent = tree.get(pair.getLeft());
+				if (parent != null)
+				{
+					Node node = pair.getRight();
+					node.parent = parent;
+					node.depth = parent.depth + 1;
+					parent.children.add(node);
+					parent.childCount++;
+	
+					node.inSchema = false;
+					node.isExplicit = false;
+					tree.put(node.uri, node);
 	
+					// resort list below after all candidates nodes added
+					list.add(node);
+					
+					it.remove();
+				}
+			}
+			if (cSize == candidates.size()) break; // exit outer loop if no change
+		}
+		if (!candidates.isEmpty())
```

Comment:
This is not an error state: it is supposed to happen most of the time.

Reply:
I can see this being the case if people use different schema.  But if people are using the same schema, to my thinking provisional terms would typically be parented in the right schema.  Please explain.

Reply:
No. There's a single list of provisional terms, and it applies completely/partially/not at all to each and every tree, of which there are currently about 2 dozen, but could ultimately be many hundreds. The process of figuring out what to apply to each tree is the same process as actually performing the addition, so there's no such thing as failure, there's only "not applicable".